lib/Reader/abi.cpp

```diff
@@ -125,17 +129,136 @@ void X86_64_Win64::computeSignatureInfo(bool IsManagedCallingConv,
   }
 }
 
-void X86_64_SysV::computeSignatureInfo(bool IsManagedCallingConv,
-                                       ABIType ResultType,
-                                       ArrayRef<ABIType> ArgTypes,
-                                       const DataLayout &DL,
-                                       ABIArgInfo &ResultInfo,
-                                       std::vector<ABIArgInfo> &ArgInfos) {
-  // TODO: RyuJIT does not implement the SysV ABI rules as decribed in "System V
-  //       Application Binary Interface". For now, agree and just use the Win64
-  //       rules.
-  X86_64_Win64::computeSignatureInfo(IsManagedCallingConv, ResultType, ArgTypes,
-                                     DL, ResultInfo, ArgInfos);
+ABIArgInfo X86_64_SysV::classify(LLILCJitContext &Context, const ABIType ABITy,
+                                 const DataLayout &DL,
+                                 bool IsManagedCallingConv, bool IsResult,
+                                 uint32_t &RequiredIntRegs,
+                                 uint32_t &RequiredSSERegs) {
+  Type *Ty = ABITy.getType();
+
+  if (Ty->isAggregateType() || Ty->isVectorTy()) {
+    SYSTEMV_AMD64_CORINFO_STRUCT_REG_PASSING_DESCRIPTOR Desc;
+
+    bool Ok = Context.JitInfo->getSystemVAmd64PassStructInRegisterDescriptor(
+        ABITy.getClass(), &Desc);
+    assert(Ok);
+
+    if (!Desc.passedInRegisters) {
```

Comment:
This could use some comments.  I take it the IsResult path is indicating that the caller will allocate a hidden buffer and pass a pointer in one of the argument registers (and that RequiredFooRegisters is tracking the passing of the arguments, not the returns), and the !IsResult path is indicating that the argument will be passed by value on the stack?